### PR TITLE
feat(api): add centralized API logging module

### PR DESCRIPTION
**Pull request by hannesrudolph**
## Summary

This PR adds a centralized API logging module to enable unified logging for inbound and outbound API calls across all providers.

### New Logging Module (`src/api/core/logging/`)
- **`api-logger.ts`** - Singleton `ApiLogger` service with `logRequest()`, `logResponse()`, `logError()` methods
- **`with-logging.ts`** - Generator wrapper for automatic stream metrics tracking (text length, reasoning length, tool calls, usage tokens)
- **`http-interceptor.ts`** - HTTP fetch wrapper that logs raw requests/responses as native objects (not stringified JSON)
- **`env-config.ts`** - Reads `ROO_CODE_LOGGING` from workspace `.env.local` file with 5s cache
- **`types.ts`** - Type definitions for logging context and metrics
- **`index.ts`** - Public exports

### Provider Integration
- Modified 20+ providers to support logging via `BaseProvider.getLogContext()`
- Added `protected get providerName()` getter pattern to all provider classes
- Integrated `createLoggingFetch()` into OpenRouter provider as a reference implementation

### Configuration
- Logging controlled by `.env.local` file in workspace: `ROO_CODE_LOGGING=true`
- Fallback to `process.env.ROO_CODE_LOGGING` (loadable via launch.json `envFile`)
- Removed hardcoded logging env vars from `launch.json` - now uses `envFile` reference

### Logs Include
- Provider name, model, operation type
- Request: message count, has tools, stream flag
- Response: text length, reasoning length, tool call count, token usage (input/output/cache/reasoning)
- Error: message, status code
- Latency timing

## Test Coverage
- 67 tests passing across the logging module and OpenRouter provider tests
- Tests cover request/response logging, stream metrics tracking, error handling, edge cases

## How to Enable
1. Create `.env.local` in workspace root with `ROO_CODE_LOGGING=true`
2. Alternatively, add `ROO_CODE_LOGGING=true` to your launch.json's `envFile`

## Files Changed
- `.vscode/launch.json` - Removed hardcoded env vars, kept envFile reference
- `src/api/core/logging/` - New module (8 files + tests)
- `src/api/providers/` - 20+ provider files updated with providerName getter
- `src/utils/logging/index.ts` - Re-exports from core logging module

----

> [!IMPORTANT]
> Introduces a centralized API logging module for unified logging across providers, with new components, provider integration, and configuration updates.
> 
>   - **Logging Module**:
>     - Adds `ApiLogger` in `api-logger.ts` with methods for logging requests, responses, and errors.
>     - Introduces `withLogging` in `with-logging.ts` for automatic logging in generator functions.
>     - Implements `createLoggingFetch` in `http-interceptor.ts` for logging HTTP requests and responses.
>     - Reads logging configuration from `.env.local` in `env-config.ts`.
>     - Defines types in `types.ts` for logging context and metrics.
>   - **Provider Integration**:
>     - Updates 20+ providers to use `ApiLogger` and `createLoggingFetch` for logging.
>     - Adds `providerName` getter to provider classes for logging context.
>     - Integrates logging into `OpenRouterHandler` as a reference implementation.
>   - **Configuration**:
>     - Logging enabled via `.env.local` with `ROO_CODE_LOGGING=true`.
>     - Removes hardcoded logging env vars from `launch.json`, uses `envFile`.
>   - **Logs Include**:
>     - Provider name, model, operation type.
>     - Request: message count, tools, stream flag.
>     - Response: text length, reasoning length, tool call count, token usage.
>     - Error: message, status code.
>     - Latency timing.
>   - **Test Coverage**:
>     - 67 tests added for logging module and OpenRouter provider.
>     - Tests cover request/response logging, metrics tracking, error handling.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for a54ba5a86efe3084b35ae2a03484dd01c3acd8fa. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

**Comment by ghost** [comment]
> **Oroocle**  Clock &nbsp;&nbsp;[See task on Roo Cloud](https://app.roocode.com/cloud-jobs/BZI6lA3W?utm_source=github-comment&utm_medium=link&utm_campaign=github.pr.review.sync)

Re-review complete. No new issues found since the last review; 2 previously flagged items remain.

- [ ] Unify / maintain compatibility for the API logging env var name (ROO_CODE_API_LOGGING vs ROO_CODE_LOGGING) and `VSCODE_DEBUG_MODE` gating
- [x] `withLogging()` should count `tool_call_partial` chunks so tool call metrics are not undercounted
- [x] `createLoggingFetch()` should avoid misleading `body.toString()` logging for non-string request bodies
- [ ] `withScopedFetchLogging()` uses global mutable `fetch`; consider guarding against concurrent (non-nested) usage so restores cannot race

> <details>
> <summary>Previous reviews</summary>
> 
> - 87b2020: [Review #1](https://app.roocode.com/cloud-jobs/qNGag3vm?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 3915224: [Review #2](https://app.roocode.com/cloud-jobs/YctCPOKB?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 2c1e46d: [Review #3](https://app.roocode.com/cloud-jobs/qNGNg3vm?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - f2b311c: [Review #4](https://app.roocode.com/cloud-jobs/8L9DaBrP?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 0a39600: [Review #5](https://app.roocode.com/cloud-jobs/3rdLF6XD?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 5ccd0f7: [Review #6](https://app.roocode.com/cloud-jobs/rJkb805j?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 3d43d34: [Review #7](https://app.roocode.com/cloud-jobs/RuIw0fMB?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 1ebacd6: [Review #8](https://app.roocode.com/cloud-jobs/UjelYszT?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - 752c29b: [Review #9](https://app.roocode.com/cloud-jobs/fGpwsFuG?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> - a7bc265: [Review #10](https://app.roocode.com/cloud-jobs/wCf9q3uN?utm_source=previous-reviews&utm_medium=link&utm_campaign=github.pr.review)
> </details>
<sub>Mention @roomote in a comment to request specific changes to this pull request or [fix all unresolved issues](https://app.roocode.com/fix-issues?repo=RooCodeInc/Roo-Code&prNumber=10176&commentId=3668206323).</sub>

File: src/api/core/logging/with-logging.ts [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+			// Track metrics based on chunk type
+			switch (chunk.type) {
+				case "text":
+					textLength += chunk.text.length
+					break
+				case "reasoning":
+					reasoningLength += chunk.text.length
+					break
+				case "tool_call":
+				case "tool_call_start":
```
Comment (ghost) [inline src/api/core/logging/with-logging.ts:64]: `withLogging()` increments `toolCallCount` for `"tool_call"`/`"tool_call_start"`, but most providers emit `"tool_call_partial"` chunks (later expanded by `NativeToolCallParser`). This makes tool-call metrics consistently undercount (often 0) even when tool calls are present.

```suggestion
				case "tool_call":
				case "tool_call_start":
				case "tool_call_partial":
					toolCallCount++
					break
```

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=10176&fixId=085ec7dcd8) or mention @roomote and request a fix.</sub>

File: src/api/core/logging/http-interceptor.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+			sanitized[key] = value
+		}
+	}
+	return sanitized
+}
+
+/**
+ * Parse body as JSON if possible, otherwise return as-is
+ */
+function parseBodyIfJson(body: BodyInit | string): unknown {
```
Comment (ghost) [inline src/api/core/logging/http-interceptor.ts]: `createLoggingFetch()` calls `parseBodyIfJson(init.body)` but `RequestInit.body` can be `Blob`/`ArrayBuffer`/`FormData`/`URLSearchParams`; `toString()` on those is usually not the actual payload. This makes the logged `body` misleading (and can be expensive). Consider only parsing when `typeof body === "string"` or when `body` is `URLSearchParams`, and otherwise log a placeholder like `[non-string body]`.

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=10176&fixId=78ceafcb0f) or mention @roomote and request a fix.</sub>

File: src/api/core/logging/env-config.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+ */
+export function isLoggingEnabled(): boolean {
+	// Check workspace .env.local first (user's current workspace)
+	const envLocal = getEnvLocalValues()
+	if (envLocal["ROO_CODE_API_LOGGING"] === "true") {
+		return true
+	}
+
+	// Fallback to process.env (populated from extension's .env.local via launch.json envFile)
+	return process.env.ROO_CODE_API_LOGGING === "true"
```
Comment (ghost) [inline src/api/core/logging/env-config.ts:103]: `isLoggingEnabled()` now checks `ROO_CODE_API_LOGGING`, but other parts of the PR (e.g. docs/comments) still refer to `ROO_CODE_LOGGING`, and `.vscode/launch.json` sets `VSCODE_DEBUG_MODE` even though it is not considered here. This can make API logging appear broken depending on which env var the user sets.

```suggestion
	if (envLocal["ROO_CODE_API_LOGGING"] === "true" || envLocal["ROO_CODE_LOGGING"] === "true") {
		return true
	}

	// Fallback to process.env (populated from extension's .env.local via launch.json envFile)
	return (
		process.env.ROO_CODE_API_LOGGING === "true" ||
		process.env.ROO_CODE_LOGGING === "true" ||
		process.env.VSCODE_DEBUG_MODE === "true"
	)
```

<sub>[Fix it with Roo Code](https://app.roocode.com/fix-issue/?repo=RooCodeInc/Roo-Code&prNumber=10176&fixId=6a473ad236) or mention @roomote and request a fix.</sub>

**Comment by daniel-lxs** [comment]
Two observations:

1. The requestTimestamps Map in api-logger.ts stores request start times keyed by requestId. These get cleaned up in logResponse() and logError(), but if a request fails before reaching those methods (e.g., network timeout at a lower level, uncaught exception), timestamps won't be cleaned up. Consider adding a TTL-based cleanup or using a WeakMap pattern if requestIds could be object references.

2. The createLoggingFetch() integration is currently only wired into OpenRouter. Is that intentional as a reference implementation, or should it be integrated into other providers that use the OpenAI client? If the latter, the inconsistency means some providers will have HTTP-level logging and others won't.
